packages/block-library/src/cover/index.php

```diff
@@ -60,10 +60,13 @@ class="wp-block-cover__image-background"
 			esc_attr( $object_position )
 		);
 
-		$content = str_replace(
-			'</span><div',
-			'</span>' . $image . '<div',
-			$content
+		/* Inserts the featured image between the (1st) cover 'background' `span` and 'inner_container' `div`,
+		   and removes eventual withespace characters between the two (typically introduced at template level) */
+		$content = preg_replace(
+		    '/(<span\b[^>]*wp-block-cover__background[\s|"][^>]*>.*<\/span>)\s*(<div\b[^>]*wp-block-cover__inner-container[\s|"][^>]*>)/U',
```

Comment:
Looking at the regex again, do you think the `span` tag matching can be removed? Doesn't seem needed as long as the opening `div` is matched. Even if there a possibility to have the div without a span before it, thinking it would be good to insert the featured image before that div. This also removes the need to use `.*` and match white space between the nodes.

So the regex would be something like: `'/<div\b[^>]+wp-block-cover__inner-container[\s|"][^>]*>/U'`.
(Note I also changed `*` to `+` after the first `[^>]` as it has to match at least one char.)

Reply:
To me the logic of the block is:
the image goes between a `span` with class "wp-block-cover__background" followed by a div with class "wp-block-cover__inner-container". If you remove one or the other, no image. What if I want to put something inside the span which in the react code is empty? Good luck!

To you and @[dmsnell](https://github.com/dmsnell) since regexes are problematic for HTML parsing, and for a better maintainability, the php porting of the react logic should be reduced to: the featured image goes before a div with class "wp-block-cover__inner-container". Do what you want before that div and good luck! 

Or , dsmell variant: if we want to require both the `span` and the `div`, at that point we must require an empty span.

Now, just in case for @draganescu and @dmsnell  `'/<div\b[^>]+wp-block-cover__inner-container[\s|"][^>]*>/U'` by  @azaozz is acceptable, and given that 

```php
$inner_container_start = // above regex;

if ( 1 !== preg_match( $inner_container_start, $content, $matches, PREG_OFFSET_CAPTURE ) {
    // handle this, we can't find our div, maybe another plugin changed it
    return $content;
}

list( $fullmatch, $offset ) = $matches[0];
$content = substr( $content, 0, $offset ) . $image_tag . substr( $content, $offset );
```
shuold be acceptable by @dmsnell 
I'm sending the 4th PR to trim the coverblock (in a few hours)
(Sorry if a wasted your time)

Reply:
@navigatrum no waster time at all, quite the reverse it will be a fruitful conversation :)

Reply:
this all seems good, though I should rephrase, as I never wanted to suggest we try and trap the `<span>`. here's the difference:

 - finding the start of the `<div>` with a RegExp is normal and doable and _not problematic_ from a parsing standpoint. this is because _it isn't parsing HTML_ at that point. it's only parsing HTML tokens, or _lexing_ the HTML.
 - trying to identify the end of a specific `<span>` before the `<div>` _is_ parsing HTML and that's the problematic part with the RegExp and that's the kind of thing we want to avoid (and it's almost always the case to be suspicious when we find ourselves adding `.*`)

seems like a nuance or a small thing, but there's a profound impact. my comments were more like, "if you are going to match the `<span>` too here's what you should be aware of).

I still don't understand why we don't want to match on `class`, but that's not a big issue to me since `wp-block-cover__inner-container` is unlikely to appear outside of where we expect it. note though, this does match a few quirky possibilities:

 - `<div title="Demonstrating how to use the wp-block-cover__inner-container styles">`
 - `<div class="x-disregard-wp-block-cover__inner-container">`

<details><summary>Feel free to ignore matching these, but if you are curious…</summary>

```
"~<div [^>]*\bclass=(\"|')((?!\1).)*\b{$CLASS}\b((?!\1).)*\1( |>)~"
```

This is almost the same as the one you gave above in your last comment. The main differences are that this:
 - specifically looks for the `class` attribute
 - makes sure we're only examining the content inside the `class` attribute by matching on the leading quote and trailing quote, ensuring they are the same, and ensuring we don't scan past the matching one with `((?!\1).)*` though I think we could replace that with the simpler `[^'"]*` since I think quotes aren't allowed as class names.
 - requires that the class name we're searching for is not a part of a bigger string through the surrounding `\b` boundary markers (that's wrong, should be `[\"' ]` since the hyphen counts as a word boundary)
 - while we're not matching a closing HTML tag I'm still using `~` instead of `/` as the PCRE pattern delimiters so we wouldn't have to escape `/` as `\/`

with the simplification that would be this instead

```
"~<div [^>]*\bclass=([\"'])[^\"']*[\"' ]wp-block-cover__inner-container[\"' ][^\"']*\1~"
```

this is simpler than general HTML attribute parsing because of the rules of CSS class names. hopefully that fulfills your curiosity. feel free to use this or ask questions or ignore it.